refactor(dogfood): return cons key in `Validator`

MaxMustermann2 opened the pull request:
Until `x/dogfood` was made compatible with `x/evm` and `x/oracle`, a `nil` validator was being returned by `ValidatorByConsAddr`. This validator was correctly processed by the `x/evidence` module when it handled equivocation evidence.

However, with the changes introduced for compatibiltiy with `x/evm`, the validator returned is no longer nil. Such a validator will have an operator address, whose presence will trigger a call to `Validator`, which is implemented by this change.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced new hooks `AfterValidatorRemoved` and `AfterValidatorCreated` for improved validator lifecycle management.

- **Improvements**
  - Enhanced validator addition process to include reverse lookup from consensus address to public key.
  - Improved error handling and context management in validator changes.
  - Updated validator retrieval logic to ensure consistency and accuracy.

- **Refactor**
  - Renamed several functions for better clarity and alignment with the codebase (`GetExocoreValidator`, `SetExocoreValidator`, etc.).

- **Bug Fixes**
  - Corrected the function return value to handle conditions more appropriately.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MaxMustermann2 committed:
refactor(dogfood): return cons key in `Validator`

Until `x/dogfood` was made compatible with `x/evm` and `x/oracle`, a
`nil` validator was being returned by `ValidatorByConsAddr`. This
validator was correctly processed by the `x/evidence` module when it
handled equivocation evidence.

However, with the changes introduced for compatibiltiy with `x/evm`, the
validator returned is no longer nil. Such a validator will have an
operator address, whose presence will trigger a call to `Validator`,
which is implemented by this change.

MaxMustermann2 committed:
Merge develop

MaxMustermann2 committed:
fix(dogfood): add expected keeper interface

MaxMustermann2 committed:
fix(dogfood): return `validator` not `val`

MaxMustermann2 commented:
Without this change, slashing is not applied, that is, the line `k.Logger(ctx).Info("slash occurs", addr, infractionHeight, slashFactor, infraction)` is never executed.

However, since slashing implementation is undergoing a redesign and currently doesn't do anything anyway, this PR is not urgent to merge.

TimmyExogenous reviewed (approved):
LGTM

bwhour commented on x/dogfood/types/hooks.go:
```diff
@@ -31,3 +31,30 @@ func (hooks MultiDogfoodHooks) AfterValidatorBonded(
 }
+
+// AfterValidatorRemoved is the implementation of types.DogfoodHooks for MultiDogfoodHooks.
+func (hooks MultiDogfoodHooks) AfterValidatorRemoved(
```
The logic of the three(AfterValidatorBonded、AfterValidatorRemoved,AfterValidatorCreated ) functions is very similar and common parts can be extracted to one function as following.

> ```
> func (hooks MultiDogfoodHooks) executeHooks(execute func(DogfoodHooks) error) error {
>     for _, hook := range hooks {
>         if err := execute(hook); err != nil {
>             return err
>         }
>     }
>     return nil
> }

```

bwhour reviewed (approved):
it looks good.

cloud8little commented:
Test object: https://github.com/ExocoreNetwork/exocore/pull/72/commits/6e032fbb3137d46acab3cc1412bee2827aa09ce3

1. Start single validator by running local_node.sh initiated with `5000 `voting power (exo18cggcpvwspnd5c6ny8wrqxpffj5zmhklprtnph).
2. Run a external node and connect to the validator (exo1u8yx7e02wr2yvs5zk3zhvehhekh4vzxdr6sv2x).
3. Register the node as an operator  and opt-into-avs  and also set consensus key.
4. Deposit and delegate `1000 `tokens to the new registered operator.
5. Check the voting power via these two rpc (the new operator voting power is correct till now): 

```
curl --request GET --url 'http://localhost:26657/validators?height=1354&page=1&per_page=30'
exocored q tendermint-validator-set
```
6. Deposit and delegate 2000 tokens to the dogfood validator, wait for next epoch, the voting power for the dogfood is still `5000`.
```
exocored query delegation QueryDelegationInfo 0xc6E1c84c2Fdc8EF1747512Cda73AaC7d338906ac_0x65 0xdAC17F958D2ee523a2206206994597C13D831ec7_0x65
delegation_infos:
  exo18cggcpvwspnd5c6ny8wrqxpffj5zmhklprtnph:
    undelegatable_share: "2000.000000000000000000"
    wait_undelegation_amount: "0"
  exo1u8yx7e02wr2yvs5zk3zhvehhekh4vzxdr6sv2x:
    undelegatable_share: "1000.000000000000000000"
    wait_undelegation_amount: "0"
```

```
exocored query delegation QueryDelegationInfo 0x3e108c058e8066da635321dc3018294ca82ddedf_0x65 0xdAC17F958D2e
e523a2206206994597C13D831ec7_0x65
delegation_infos:
  exo18cggcpvwspnd5c6ny8wrqxpffj5zmhklprtnph:
    undelegatable_share: "5000.000000000000000000"
    wait_undelegation_amount: "0"
```

```
exocored q tendermint-validator-set
block_height: "1354"
total: "2"
validators:
- address: exovalcons18z3p42xn8pjk338upvzp794h02wh7p4t7jj9jx
  proposer_priority: "375"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: 8PaRnlIsW5fbLIJVv/dD+d/d162fw3ywwWcLSA0PmRQ=
  voting_power: "5000"
- address: exovalcons1fllfq6nj5tc9wu8a82dtqd8axfxva48wzna4pa
  proposer_priority: "-375"
  pub_key:
    type: tendermint/PubKeyEd25519
    value: XLtFCK0/nB1xExSXEhH5kaxRte3aIXSGaBfWSeNOtpE=
  voting_power: "1000"	
```

but tried to export the genesis with https://github.com/ExocoreNetwork/exocore/pull/95/commits/e983c46aac8327d014421ffdcb497f0d40b90d28 and checked the dogfood field data: 

 ```
"dogfood": {
      "consensus_addrs_to_prune": [],
      "last_total_power": "8000",
      "opt_out_expiries": [],
      "params": {
        "asset_ids": [
          "0xdac17f958d2ee523a2206206994597c13d831ec7_0x65"
        ],
        "epoch_identifier": "hour",
        "epochs_until_unbonded": 7,
        "historical_entries": 10000,
        "max_validators": 100
      },
      "undelegation_maturities": [],
      "val_set": [
        {
          "power": "7000",
          "public_key": "0xf0f6919e522c5b97db2c8255bff743f9dfddd7ad9fc37cb0c1670b480d0f9914"
        },
        {
          "power": "1000",
          "public_key": "0x5cbb4508ad3f9c1d711314971211f991ac51b5edda2174866817d649e34eb691"
        }
      ]
    }
```

MaxMustermann2 committed:
Merge develop

MaxMustermann2 committed:
Merge develop

MaxMustermann2 committed:
fix(operator,dogfood): return val with key

MaxMustermann2 committed:
fix(dogfood): correctly set bonded status

The validators loaded from x/dogfood are considered bonded, while the
validators pulled from x/operator can have any status.

MaxMustermann2 commented on x/dogfood/keeper/validators.go:
```diff
@@ -53,16 +54,43 @@
-		val, found := k.GetValidator(ctx, addr)
+		val, found := k.GetExocoreValidator(ctx, addr)
 		switch found {
 		case true:
 			// update or delete an existing validator.
 			// assumption: power can not be negative.
 			if change.Power < 1 {
-				k.DeleteValidator(ctx, addr)
+				// guard for errors within the hooks.
+				cc, writeFunc := ctx.CacheContext()
+				k.DeleteExocoreValidator(cc, addr)
+				// sdk slashing.AfterValidatorRemoved deletes the lookup from cons address to
+				// cons pub key
+				err = k.Hooks().AfterValidatorRemoved(cc, sdk.ConsAddress(addr), nil)
+				if err != nil {
+					continue
+				}
+				writeFunc()
 			} else {
 				val.Power = change.Power
-				k.SetValidator(ctx, val)
+				// guard for errors within the hooks.
+				cc, writeFunc := ctx.CacheContext()
+				k.SetExocoreValidator(ctx, val)
+				// sdk slashing.AfterValidatorCreated stores the lookup from cons address to
+				// cons pub key. it loads the validator from `valAddr` (operator address)
+				// via stakingkeeeper.Validator(ctx, valAddr)
+				// then it fetches the cons pub key from said validator to generate the lookup
+				found, accAddress := k.operatorKeeper.GetOperatorAddressForChainIDAndConsAddr(
+					ctx, ctx.ChainID(), sdk.ConsAddress(addr),
+				)
+				if !found {
+					// should never happen
+					continue
+				}
+				err = k.Hooks().AfterValidatorCreated(cc, sdk.ValAddress(accAddress))
+				if err != nil {
+					continue
+				}
+				writeFunc()
```
I don't control the interface of this function and hence it cannot return any errors.

MaxMustermann2 commented on x/dogfood/types/hooks.go:
```diff
@@ -31,3 +31,30 @@ func (hooks MultiDogfoodHooks) AfterValidatorBonded(
 	return nil
 }
+
+// AfterValidatorRemoved is the implementation of types.DogfoodHooks for MultiDogfoodHooks.
+func (hooks MultiDogfoodHooks) AfterValidatorRemoved(
+	ctx sdk.Context,
+	consAddr sdk.ConsAddress,
+	operator sdk.ValAddress,
+) error {
+	for _, hook := range hooks {
+		if err := hook.AfterValidatorRemoved(ctx, consAddr, operator); err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
This is not readable. I would rather retain the original.

MaxMustermann2 committed:
doc(dogfood): add error logs

MaxMustermann2 committed:
fix(ci): specify go version in consensuswarn

MaxMustermann2 commented:
The failure from `consensuswarn` will be fixed when this PR is merged. Since the workflow runs on the base `develop` branch and not the PR branch, it will wrongly appear to fail within this PR.

Here are logs from a local run. I used the file `asd.json` to specify the PR number. The workflow was run with `remote.origin.url` pointing to this repo and not my fork.

```
$ act --workflows .github/workflows/consensuswarn.yml --eventpath asd.json pull_request_target
INFO[0000] Using docker host 'unix:///var/run/docker.sock', and daemon socket 'unix:///var/run/docker.sock' 
[Consensus Warn/main] 🚀  Start image=catthehacker/ubuntu:act-lates
[Consensus Warn/main]   🐳  docker pull image=catthehacker/ubuntu:act-latest platform= username= forcePull=tru
[Consensus Warn/main]   🐳  docker create image=catthehacker/ubuntu:act-latest platform= entrypoint=["tail" "-f" "/dev/null"] cmd=[] network="host
[Consensus Warn/main]   🐳  docker run image=catthehacker/ubuntu:act-latest platform= entrypoint=["tail" "-f" "/dev/null"] cmd=[] network="host
[Consensus Warn/main]   ☁  git clone 'https://github.com/orijtech/consensuswarn' # ref=956f047a43f56021a28afdfb2a2291a20955f48d
[Consensus Warn/main] ⭐ Run Main actions/checkout@v
[Consensus Warn/main]   🐳  docker cp src=/home/user/go/src/github.com/ExocoreNetwork/exocore/. dst=/home/user/go/src/github.com/ExocoreNetwork/exocor
[Consensus Warn/main]   ✅  Success - Main actions/checkout@v
[Consensus Warn/main] ⭐ Run Main orijtech/consensuswarn@956f047a43f56021a28afdfb2a2291a20955f48
[Consensus Warn/main]   🐳  docker build -t act-orijtech-consensuswarn-956f047a43f56021a28afdfb2a2291a20955f48d-dockeraction:latest /home/user/.cache/act/orijtech-consensuswarn@956f047a43f56021a28afdfb2a2291a2
955f48d/
[Consensus Warn/main]   🐳  docker pull image=act-orijtech-consensuswarn-956f047a43f56021a28afdfb2a2291a20955f48d-dockeraction:latest platform= username= forcePull=fals
[Consensus Warn/main]   🐳  docker create image=act-orijtech-consensuswarn-956f047a43f56021a28afdfb2a2291a20955f48d-dockeraction:latest platform= entrypoint=[] cmd=["-ghtoken" "" "-apiurl" "https://api.github.
om" "-repository" "ExocoreNetwork/exocore" "-pr" "72" "-roots" "github.com/ExocoreNetwork/exocore/app.ExocoreApp.DeliverTx,github.com/ExocoreNetwork/exocore/app.ExocoreApp.BeginBlocker,github.com/ExocoreNetwork
/exocore/app.ExocoreApp.EndBlocker"] network="container:act-Consensus-Warn-main-8da93d63791b26a1daaf75d558bed1956bf209a956bb8c24372d7cbc6d35372b"
[Consensus Warn/main]   🐳  docker run image=act-orijtech-consensuswarn-956f047a43f56021a28afdfb2a2291a20955f48d-dockeraction:latest platform= entrypoint=[] cmd=["-ghtoken" "" "-apiurl" "https://api.github.com
 "-repository" "ExocoreNetwork/exocore" "-pr" "72" "-roots" "github.com/ExocoreNetwork/exocore/app.ExocoreApp.DeliverTx,github.com/ExocoreNetwork/exocore/app.ExocoreApp.BeginBlocker,github.com/ExocoreNetwork/ex
ocore/app.ExocoreApp.EndBlocker"] network="container:act-Consensus-Warn-main-8da93d63791b26a1daaf75d558bed1956bf209a956bb8c24372d7cbc6d35372b"
[Consensus Warn/main]   ✅  Success - Main orijtech/consensuswarn@956f047a43f56021a28afdfb2a2291a20955f48
[Consensus Warn/main] Cleaning up container for job main
[Consensus Warn/main] 🏁  Job succeeded
```

cloud8little commented:
> Test object: [6e032fb](https://github.com/ExocoreNetwork/exocore/commit/6e032fbb3137d46acab3cc1412bee2827aa09ce3)
> 
> 1. Start single validator by running local_node.sh initiated with `5000 `voting power (exo18cggcpvwspnd5c6ny8wrqxpffj5zmhklprtnph).
> 2. Run a external node and connect to the validator (exo1u8yx7e02wr2yvs5zk3zhvehhekh4vzxdr6sv2x).
> 3. Register the node as an operator  and opt-into-avs  and also set consensus key.
> 4. Deposit and delegate `1000 `tokens to the new registered operator.
> 5. Check the voting power via these two rpc (the new operator voting power is correct till now):
> 
> ```
> curl --request GET --url 'http://localhost:26657/validators?height=1354&page=1&per_page=30'
> exocored q tendermint-validator-set
> ```
> 
> 6. Deposit and delegate 2000 tokens to the dogfood validator, wait for next epoch, the voting power for the dogfood is still `5000`.
> 


The above issue does not happen after https://github.com/ExocoreNetwork/exocore/pull/98 was merged.

cloud8little commented:
Test object: https://github.com/ExocoreNetwork/exocore/pull/72/commits/ccec0bbe3afdb057886281772c0fb2726fcf0205

Test passed, when operator node get slashed because of downtime, the slash log shows: 

```
8:58AM INF ending epoch identifier=minute module=x/epochs number=18
8:58AM INF slash occurs module=x/operator
8:58AM INF slashing and jailing validator due to liveness fault height=301 jailed_until=2024-06-14T09:08:44Z min_height=300 module=x/slashing slashed=0.010000000000000000 threshold=50 validator=exovalcons1fllf
q6nj5tc9wu8a82dtqd8axfxva48wzna4pa
8:58AM INF validator set changed, force seal all active rounds module=x/oracle
```

MaxMustermann2 committed:
Merge develop
